File: code/design-patterns/behavioral/visitor/go/visitor.go
package main

import (
	"fmt"
	"math"
)

// Visitor interface
type Visitor interface {
	VisitCircle(c *Circle)
	VisitSquare(s *Square)
	VisitTriangle(t *Triangle)
}

// Concrete Visitor for calculating area
type AreaCalculator struct{}

func (a *AreaCalculator) VisitCircle(c *Circle) {
	area := math.Pi * c.Radius * c.Radius
	fmt.Printf("Circle area: %.2f\n", area)
}

func (a *AreaCalculator) VisitSquare(s *Square) {
	area := s.Side * s.Side
	fmt.Printf("Square area: %.2f\n", area)
}

func (a *AreaCalculator) VisitTriangle(t *Triangle) {
	area := 0.5 * t.Base * t.Height
	fmt.Printf("Triangle area: %.2f\n", area)
}

// Concrete Visitor for calculating perimeter
type PerimeterCalculator struct{}

func (p *PerimeterCalculator) VisitCircle(c *Circle) {
	perimeter := 2 * math.Pi * c.Radius
	fmt.Printf("Circle perimeter: %.2f\n", perimeter)
}

func (p *PerimeterCalculator) VisitSquare(s *Square) {
	perimeter := 4 * s.Side
	fmt.Printf("Square perimeter: %.2f\n", perimeter)
}

func (p *PerimeterCalculator) VisitTriangle(t *Triangle) {
	perimeter := t.SideA + t.SideB + t.SideC
	fmt.Printf("Triangle perimeter: %.2f\n", perimeter)
}

// Shape interface
type Shape interface {
	Accept(visitor Visitor)
}

// Circle struct
type Circle struct {
	Radius float64
}

func (c *Circle) Accept(visitor Visitor) {
	visitor.VisitCircle(c)
}

// Square struct
type Square struct {
	Side float64
}

func (s *Square) Accept(visitor Visitor) {
	visitor.VisitSquare(s)
}

// Triangle struct
type Triangle struct {
	SideA  float64
	SideB  float64
	SideC  float64
	Base   float64
	Height float64
}

func NewTriangle(sideA, sideB, sideC float64) *Triangle {
	// assuming it's an isosceles triangle for simplicity
	height := math.Sqrt(sideB*sideB - (0.5*sideA)*(0.5*sideA))
	return &Triangle{
		SideA:  sideA,
		SideB:  sideB,
		SideC:  sideC,
		Base:   sideA,
		Height: height,
	}
}

func (t *Triangle) Accept(visitor Visitor) {
	visitor.VisitTriangle(t)
}

File: code/design-patterns/behavioral/visitor/python/visitor.py
from abc import ABC, abstractmethod


# Visitor interface
class Visitor(ABC):
    @abstractmethod
    def visit_circle(self, circle):
        pass

    @abstractmethod
    def visit_square(self, square):
        pass

    @abstractmethod
    def visit_triangle(self, triangle):
        pass


# Concrete Visitor for calculating area
class AreaCalculator(Visitor):
    def visit_circle(self, circle):
        area = 3.14 * (circle.radius**2)
        print(f"Circle area: {area}")

    def visit_square(self, square):
        area = square.side * square.side
        print(f"Square area: {area}")

    def visit_triangle(self, triangle):
        area = 0.5 * triangle.base * triangle.height
        print(f"Triangle area: {area}")


# Concrete Visitor for calculating perimeter
class PerimeterCalculator(Visitor):
    def visit_circle(self, circle):
        perimeter = 2 * 3.14 * circle.radius
        print(f"Circle perimeter: {perimeter}")

    def visit_square(self, square):
        perimeter = 4 * square.side
        print(f"Square perimeter: {perimeter}")

    def visit_triangle(self, triangle):
        perimeter = triangle.side_a + triangle.side_b + triangle.side_c
        print(f"Triangle perimeter: {perimeter}")


# Shape interface
class Shape(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass


# Concrete Shape: Circle
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def accept(self, visitor):
        visitor.visit_circle(self)


# Concrete Shape: Square
class Square(Shape):
    def __init__(self, side):
        self.side = side

    def accept(self, visitor):
        visitor.visit_square(self)


# Concrete Shape: Triangle
class Triangle(Shape):
    def __init__(self, side_a, side_b, side_c):
        self.side_a = side_a
        self.side_b = side_b
        self.side_c = side_c
        self.base = side_a
        self.height = (
            side_b**2 - (0.5 * side_a) ** 2
        ) ** 0.5  # assuming it's an isosceles triangle

    def accept(self, visitor):
        visitor.visit_triangle(self)
